### PR TITLE
[AUTO] Fix the mismatch issue between AUTO tensor shape and HW tensor shape

### PR DESCRIPTION
**Pull request by yangwang201911**
### Details:
 - Correct tensor shape of AUTO before setting it to selected HW infer request.
 - Remove CPU shape.compatible checks when setting output tensor with dynamic shape

### Tickets:
 - CVS-136756

### CODE REVIEW

File: src/plugins/auto/src/infer_request.cpp [hunk 1]
```diff
@@ -97,8 +97,12 @@ void ov::auto_plugin::InferRequest::set_tensors_to_another_request(const SoAsync
         auto type = tensor->get_element_type();
         bool is_remote  = std::dynamic_pointer_cast<ov::IRemoteTensor>(tensor._ptr) ||
             std::dynamic_pointer_cast<ov::IRemoteTensor>(req->get_tensor(it)._ptr);
-        if (is_remote || req->get_tensor(it)->data(type) != tensor->data(type))
+        if (is_remote || req->get_tensor(it)->data(type) != tensor->data(type)) {
+            // Update tensor shape if it differs from the selected infer request tensor shape
```
Comment (songbell) [inline src/plugins/auto/src/infer_request.cpp:101]:
```suggestion
            // temp workaround for NMS-like operations been converted to static shape with upper bound in gpu plugin
```

File: src/plugins/auto/src/infer_request.cpp [hunk 2]
```diff
@@ -97,8 +97,12 @@ void ov::auto_plugin::InferRequest::set_tensors_to_another_request(const SoAsync
...
+            if (req->get_tensor(it)->get_shape() != tensor->get_shape())
```
Comment (songbell) [inline src/plugins/auto/src/infer_request.cpp:102]: let's add more strict check, only check the tensor is dynamic, and the tensor in hardware is not
and let's enable a test for it

Comment (yangwang201911) [reply src/plugins/auto/src/infer_request.cpp]: Updated. will enable test case soon.

File: src/plugins/auto/src/infer_request.cpp [hunk 3]
```diff
@@ -97,8 +97,12 @@ void ov::auto_plugin::InferRequest::set_tensors_to_another_request(const SoAsync
         auto type = tensor->get_element_type();
         bool is_remote  = std::dynamic_pointer_cast<ov::IRemoteTensor>(tensor._ptr) ||
             std::dynamic_pointer_cast<ov::IRemoteTensor>(req->get_tensor(it)._ptr);
-        if (is_remote || req->get_tensor(it)->data(type) != tensor->data(type))
+        if (is_remote || req->get_tensor(it)->data(type) != tensor->data(type)) {
+            // temp workaround for NMS-like operations been converted to static shape with upper bound in gpu plugin
+            if (it.get_partial_shape().is_dynamic() && req->get_tensor(it)->get_shape() != tensor->get_shape())
```
Comment (songbell) [inline src/plugins/auto/src/infer_request.cpp:102]: maybe need to careful about != here, can you check if cpu can hit this "if"

Comment (yangwang201911) [reply src/plugins/auto/src/infer_request.cpp]: remove shape incomputable checking logic of output tensor setting for CPU plugin with dynamic shape. Also add corresponding test cases.

File: src/plugins/auto/src/infer_request.cpp [hunk 4]
```diff
@@ -97,8 +97,12 @@ void ov::auto_plugin::InferRequest::set_tensors_to_another_request(const SoAsync
         auto type = tensor->get_element_type();
         bool is_remote  = std::dynamic_pointer_cast<ov::IRemoteTensor>(tensor._ptr) ||
             std::dynamic_pointer_cast<ov::IRemoteTensor>(req->get_tensor(it)._ptr);
-        if (is_remote || req->get_tensor(it)->data(type) != tensor->data(type))
+        if (is_remote || req->get_tensor(it)->data(type) != tensor->data(type)) {
+            // temp workaround for NMS-like operations been converted to static shape with upper bound in gpu plugin
+            if (it.get_partial_shape().is_dynamic() && req->get_tensor(it)->get_size() != 0)
```
Comment (ilya-lavrenov) [inline src/plugins/auto/src/infer_request.cpp:102]: do we have a ticket on GPU plugin to fix behavior?

Comment (yangwang201911) [reply src/plugins/auto/src/infer_request.cpp:102]: CVS-138478

**Comment by wangleis** [comment]
@dmitry-gorokhov @slyalin Could you please review the change in CPU plugin of this PR?

**Review by wangleis — APPROVED** [review]

File: src/plugins/intel_cpu/src/infer_request.cpp [hunk 5]
```diff
@@ -419,7 +419,7 @@ void SyncInferRequest::set_tensor(const ov::Output<const ov::Node>& in_port, con
         const auto& shape = port.get_partial_shape();
         const bool isDynamic = shape.is_dynamic();
 
-        if (!shape.compatible(ov::PartialShape(tensor->get_shape()))) {
+        if (!shape.compatible(ov::PartialShape(tensor->get_shape())) && tensor->get_size() != 0) {
```
Comment (songbell) [inline src/plugins/intel_cpu/src/infer_request.cpp:422]: does it make sense that in general the hardware plugins can align the the common basic compatibility checking?  @ilya-lavrenov 
we have below in gpu plugin, which performs similar checking, but with different conditions in if clause:
https://github.com/openvinotoolkit/openvino/blob/master/src/plugins/intel_gpu/src/plugin/sync_infer_request.cpp#L144

Comment (ilya-lavrenov) [reply src/plugins/intel_cpu/src/infer_request.cpp:422]: I suppose we discussed in email thread "[CPU Plugin] Infer request failed to set customer output tensor with dynamic shape" that checks should be aligned.

**Review by dmitry-gorokhov — APPROVED** [review]
CPU plugin changes looks good to me

**Review by songbell — APPROVED** [review]

**Pull request merged**